static/app/stores/sidebarProvider.tsx

```diff
@@ -0,0 +1,62 @@
+import {createContext, useContext, useReducer} from 'react';
+
+function assertNever(x: never) {
+  throw new TypeError('Unhandled sidebar reducer action: ' + x);
+}
+
+import {SidebarPanelKey} from 'sentry/components/sidebar/types';
+
+type State = SidebarPanelKey | '';
+
+type HidePanelAction = {type: 'hide panel'};
+type TogglePanelAction = {payload: SidebarPanelKey; type: 'toggle panel'};
+type ActivatePanelAction = {payload: SidebarPanelKey; type: 'activate panel'};
+type SidebarAction = HidePanelAction | TogglePanelAction | ActivatePanelAction;
+
+const SidebarContext = createContext<[State, React.Dispatch<SidebarAction>] | null>(null);
+
+export function SidebarReducer(state: State, action: SidebarAction) {
+  switch (action.type) {
+    case 'activate panel':
+      return action.payload;
+    case 'hide panel':
+      return '';
+    case 'toggle panel':
+      if (state === action.payload) {
+        return '';
+      }
+      return action.payload;
+    default:
+      assertNever(action);
+      return state;
+  }
+}
+
+interface Props {
+  children: React.ReactNode;
+  initialPanel?: State;
+}
+
+export function SidebarProvider(props: Props) {
+  const reducer = useReducer(SidebarReducer, props.initialPanel ?? '');
+
+  return (
+    <SidebarContext.Provider value={reducer}>{props.children}</SidebarContext.Provider>
+  );
+}
+
+export function useSidebarDispatch() {
+  const context = useContext(SidebarContext);
+  if (context === null) {
+    throw new Error('useSidebarDispatch was called outside of SidebarProvider');
+  }
+  return context[1];
+}
+
+export function useSidebar() {
+  const context = useContext(SidebarContext);
+  if (context === null) {
+    throw new Error('useSidebar was called outside of SidebarProvider');
+  }
+  return context;
+}
```

Comment:
Why two? Should the second one be `return context[0]`.

IMO just a `useSidebar` is fine where it returns `[visiblePanel, dispatch]`

Reply:
Some places only require a dispatch and I didnt want to const `[_, dispatch] = useSidebar()`, either works though

Reply:
Another thing to keep in mind is that we can separate the state and dispatch into separate contexts. With a single context components consuming `useSidebarDispatch` will still rerender on state updates. Not a big deal here, but could be for more complicated state.

Next.js official example: https://github.com/vercel/next.js/blob/canary/examples/with-context-api/components/Counter.js

https://hswolff.com/blog/how-to-usecontext-with-usereducer/#performance-concerns

Reply:
Great point @malwilley - I opted for a single one here because the component is mostly scope to just the sidebar and I didnt want to optimize anything yet, but that would be a nice improvement especially because it is transparent from a consumer pov (with the tradeoff of having more boilerplate). If we merge this, I will probably simplify the store and introduce that with it.